test/e2e/util/oauth.go

```diff
@@ -0,0 +1,208 @@
+		}
+		hostedCluster.Spec.Configuration.OAuth = &v1.OAuthSpec{
+			IdentityProviders: []v1.IdentityProvider{
+				{
+					Name:          "my_htpasswd_provider",
+					MappingMethod: v1.MappingMethodClaim,
+					IdentityProviderConfig: v1.IdentityProviderConfig{
+						Type: v1.IdentityProviderTypeHTPasswd,
+						HTPasswd: &v1.HTPasswdIdentityProvider{
+							FileData: v1.SecretNameReference{
+								Name: secret.Name,
+							},
+						},
+					},
+				},
+			},
+		}
+
+		err = client.Update(ctx, hostedCluster)
+		g.Expect(err).ToNot(HaveOccurred(), "failed to update hostedcluster identity providers")
+
+		guestKubeConfigSecretData, err := WaitForGuestKubeConfig(t, ctx, client, hostedCluster)
+		g.Expect(err).NotTo(HaveOccurred(), "couldn't get kubeconfig")
+		guestConfig, err := clientcmd.RESTConfigFromKubeConfig(guestKubeConfigSecretData)
+		g.Expect(err).ToNot(HaveOccurred())
+
+		// wait for oauth route to be ready
+		oauthRoute := WaitForOAuthRouteReady(t, ctx, client, guestConfig, hostedCluster)
+
+		// wait for oauth token request to succeed
+		access_token := WaitForOAuthToken(t, ctx, oauthRoute, guestConfig, "testuser", "password")
+
+		user, err := GetUserForToken(guestConfig, access_token)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(user.Name).To(Equal("testuser"))
+	})
+}
+
+func WaitForOAuthToken(t *testing.T, ctx context.Context, oauthRoute *routev1.Route, restConfig *restclient.Config, username, password string) string {
+	g := NewWithT(t)
+
+	oauthClient := configmanifests.OAuthServerChallengingClient().Name
+	tokenReqUrl := fmt.Sprintf("https://%s/oauth/authorize?response_type=token&client_id=%s", oauthRoute.Spec.Host, oauthClient)
+	request, err := http.NewRequest(http.MethodGet, tokenReqUrl, nil)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	request.Header.Set("Authorization", getBasicHeader(username, password))
+	request.Header.Set("X-CSRF-Token", "1")
+
+	transport, err := restclient.TransportFor(restclient.AnonymousClientConfig(restConfig))
+	g.Expect(err).ToNot(HaveOccurred(), "error getting transport")
+
+	httpClient := &http.Client{Transport: transport}
+	httpClient.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		// don't resolve redirects and return the response instead
+		return http.ErrUseLastResponse
+	}
+
+	var access_token string
+	err = wait.PollImmediateWithContext(ctx, time.Second, time.Minute, func(ctx context.Context) (done bool, err error) {
+		resp, err := httpClient.Do(request)
+		if err != nil {
+			t.Logf("Waiting for OAuth token request to succeed")
+			return false, nil
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusFound {
+			t.Logf("Waiting for OAuth token request status code %v, got %v", http.StatusFound, resp.StatusCode)
+			return false, nil
+		}
+
+		// extract access_token from redirect URL
+		access_token, err = extractAccessToken(resp)
+		if err != nil {
+			t.Logf("Failed to extract access token from redirect url")
+			return false, nil
```

Comment:
shouldn't it be legit we return the err here?

Reply:
I was thinking that the oauth server could have intermittent errors and return an invalid redirect url, and a retry could solve the problem. what do you think?